### PR TITLE
[RFC] Deferred device initialization

### PR DESCRIPTION
**Pull request by edersondisouza**
Following discussions at #64869, this PR implements [Option 1](https://github.com/zephyrproject-rtos/zephyr/issues/39896#issuecomment-1767250893) from #39896.

With this PR, a new devicetree property, `zephyr,deferred-init` can be used to make a device not be initialised during boot. The device can be initialised by calling `device_init()` from the application.

Under the hood, this PR achieves that by grouping the deferred devices intialisation code in a different ELF section. This way, there's no overhead during normal initialisation: neither memory nor processing penalty. Deferred initialisation however loops over the deferred devices and initialises the matching one, incurring some processing penalty, which is hopefully acceptable.

A new test helps showcase the new feature.

### Opens

 - Dependencies are not handled. If a non-deferred device depends on a deferred one, its initialisation will simply fail. We could A) issue warnings/errors during build or B) forcefully make all dependencies also be deferred.
 - Is the processing penalty to initialise a deferred device acceptable? If not, one thing I can think is to expose the initialiser function as non-static and providing a macro to get it and call it directly (think of `DEVICE_INIT(DT_PATH(some_node))`), but this seems way too ugly. Any ideas?
 - This PR doesn't check if a deferred device is initialised more than once - it will simply re-run the code. My reasoning is that Zephyr currently marks a device as initialised if initialisation was _attempted_, not if it _succeeded_. This behaviour is kept for deferred devices for the sake of consistency. But if someone wants to retry initialise some device, checking for attempted initialisation would prevent that.
 - This PR doesn't attempt to tackle the "device de-init" problem - only selective device init.

### CODE REVIEW

File: include/zephyr/device.h [hunk 1]
```diff
@@ -149,6 +149,8 @@ typedef int16_t device_handle_t;
 #define DEVICE_DT_NAME(node_id)                                                \
 	DT_PROP_OR(node_id, label, DT_NODE_FULL_NAME(node_id))
 
+#define DEVICE_DT_DEFER(node_id)                                               \
+	DT_PROP_OR(node_id, zephyr_deferred_init, false)
```
Comment (gmarull) [inline include/zephyr/device.h:153]: `DT_PROP(node_id, zephyr_deferred_init)` (boolean prop, defaults to 0 if not defined)

File: kernel/init.c [hunk 2]
```diff
@@ -280,36 +315,41 @@ static void z_sys_init_run_level(enum init_level level)
...
+
+	if (dev == NULL) {
+		return -ENOENT;
+	}
+
+	for (entry = __deferred_init_start; entry <= __deferred_init_end; entry++) {
+		if (entry->dev == dev) {
+			return do_device_init(entry);
+		}
+	}
```
Comment (gmarull) [inline kernel/init.c:338]: please use iterable sections

**Review by gmarull — CHANGES_REQUESTED** [review]

File: kernel/init.c (hunk 2 above)
Comment (cfriedt) [reply kernel/init.c:338]: Yes - I can't remember the exact macro off the top of my head `ITERABLE_SECTION_FOR_EACH()` maybe? There is likely an ALT version that should allow for pointer aliasing if necessary

**Review by andyross — DISMISSED** [review]
Clever and useful.  Dropping a +1 even though it'll probably respin for the macrobatic review comments.

**Comment by edersondisouza** [comment]
v2:
  - Use `STRUCT_SECTION_FOREACH_ALTERNATE` to loop over init entries;
  - `DT_PROP` instead of `DT_PROP_OR`;
  - A bit more docs and style fixes.

**Review by npitre — DISMISSED** [review]

File: include/zephyr/device.h [hunk 3]
```diff
@@ -149,6 +149,16 @@ typedef int16_t device_handle_t;
 #define DEVICE_DT_NAME(node_id)                                                \
 	DT_PROP_OR(node_id, label, DT_NODE_FULL_NAME(node_id))
 
+/**
+ * @brief Deterimine if a devicetree node initialization should be deferred.
```
Comment (nashif) [inline include/zephyr/device.h:153]: typo

**Comment by edersondisouza** [comment]
v3:
  - Fix typo.

**Review by npitre — DISMISSED** [review]

File: include/zephyr/linker/common-rom/common-rom-kernel-devices.ld [hunk 4]
```diff
@@ -18,6 +18,9 @@
 		CREATE_OBJ_LEVEL(init, APPLICATION)
 		CREATE_OBJ_LEVEL(init, SMP)
 		__init_end = .;
+		__deferred_init_list_start = .;
+		KEEP(*(.z_deferred_init))
+		__deferred_init_list_end = .;
```
Comment (gmarull) [inline include/zephyr/linker/common-rom/common-rom-kernel-devices.ld:23]: why isn't this use its own section using eg `ITERABLE_SECTION_ROM`?

**Comment by fabiobaltieri** [comment]
One problem I see with this approach is that once the device is marked for deferred-initialization, if the application wants to initialize it in the intended point it'd be very hard to do so because the full sequencing information is lost: I imagine on could set up a SYS_INIT at the same level as priority as the original device, but the ordinal (hence the devicetree inferred dependency) will be lost.

What if instead of marking the device as "deferred-initialization" in the tree we would have an iterable sections of callbacks that can be defined by the application to skip device initialization in the normal init loop?

That way the callback could see if the device has to be skipped or not in runtime and let it initialize in the intended point in the sequence, or skip it and initialize it late with `device_init()`, keeping in mind that that has no guarantees of priority validation from the devicetree.

**Comment by cfriedt** [comment]
+1 to @fabiobaltieri 's suggestion.

I was going to suggest the same thing, but didn't want to overcomplicate things.

It would be great if there were a cleaner interface to query if a deferred init device should be expected to initialize successfully without having to call the init function.

That way as well, categories of deferred init devices (e.g. those that only rely on the k_uptime_get()) can even share the same logic, which would save some ROM space.

The more important aspect is that it is a bit more testable.

E.g.

```cpp
typedef bool (*def_init_ok_t)(const struct device *dev);
```

Something along those lines.

**Comment by fabiobaltieri** [comment]
> I was going to suggest the same thing, but didn't want to overcomplicate things.

I think it would simplify few things actually, you would not have to do any DT changes and `do_device_init` would just go through all devices in the existing sections and I guess error out if the device is initialized already, so it'd just be a hook into `z_sys_init_run_level` and the new API.

**Comment by carlescufi** [comment]
Architecture WG:

- @andyross approves of the proposal in this PR
- @fabiobaltieri echoes his own comment [here](https://github.com/zephyrproject-rtos/zephyr/pull/67335#issuecomment-1886887611)
-

**Comment by edersondisouza** [comment]
I gave a shot at @fabiobaltieri idea at #67877.

**Comment by henrikbrixandersen** [comment]
Architecture WG meeting, 2024-01-23:
- Please provide review feedback on both this PR and the alternative implementation available in https://github.com/zephyrproject-rtos/zephyr/pull/67877.
- Concerns raised that working on multiple, different solutions for this may further delay this feature.. We need to decide on one and stick with it.

**Comment by tbursztyka** [comment]
I prefer something simpler like this PR proposes over the one with callbacks. The least work required for the device driver developer, is imo the best.

A few comments: 

> With this PR, a new devicetree property, `zephyr,deferred-init` can be used to make a device not be initialised during boot. The device can be initialised by calling `device_init()` from the application.

I am not entirely sure this is the path we want to follow. I mean, deferred init is just a "manual level", so instead of such DTS property, one could just add a MANUAL level (note that it would not be device specific then, but init specific).
This is a software property, not a hw property and I don't see this getting into DTS.

> Under the hood, this PR achieves that by grouping the deferred devices intialisation code in a different ELF section. This way, there's no overhead during normal initialisation: neither memory nor processing penalty. Deferred initialisation however loops over the deferred devices and initialises the matching one, incurring some processing penalty, which is hopefully acceptable.

If a MANUAL level would be added, it would be even simpler/faster:  CREATE_OBJ_LEVEL(init, MANUAL) in the level sections and that's it.

>     * Dependencies are not handled. If a non-deferred device depends on a deferred one, its initialisation will simply fail. We could A) issue warnings/errors during build or B) forcefully make all dependencies also be deferred.

I think to get such feature, it would need to manage dependencies. And I guess the rule would have to be : if a non-deferred device has a dependency on a deferred device... then the deferred device would no longer be deferred.


>     * This PR doesn't check if a deferred device is initialised more than once - it will simply re-run the code. My reasoning is that Zephyr currently marks a device as initialised if initialisation was _attempted_, not if it _succeeded_. This behaviour is kept for deferred devices for the sake of consistency. But if someone wants to retry initialise some device, checking for attempted initialisation would prevent that.

That has been a long-lasting issue indeed.
We would need to keep track of device runtime status. See #23589 and specifically "device: Add the preliminary function for error reporting". This is obviously an obsolete PR, but the idea is still valid.

**Comment by keith-zephyr** [comment]
> > With this PR, a new devicetree property, `zephyr,deferred-init` can be used to make a device not be initialised during boot. The device can be initialised by calling `device_init()` from the application.
> 
> I am not entirely sure this is the path we want to follow. I mean, deferred init is just a "manual level", so instead of such DTS property, one could just add a MANUAL level (note that it would not be device specific then, but init specific). This is a software property, not a hw property and I don't see this getting into DTS.
> 

I disagree that the deferred initialization is strictly a software property.  My main use case for this is to select the correct driver to initialize at runtime, which allows the same firmware image to be used on boards with multiple part stuffing options.  I also have use cases where a board needs 2 instances of a driver, but one instance should automatically be initialized (it is present on all board variants) but the 2nd instance requires runtime check (it is only used on some variants).

**Comment by tbursztyka** [comment]
> I disagree that the deferred initialization is strictly a software property. My main use case for this is to select the correct driver to initialize at runtime, which allows the same firmware image to be used on boards with multiple part stuffing options. I also have use cases where a board needs 2 instances of a driver, but one instance should automatically be initialized (it is present on all board variants) but the 2nd instance requires runtime check (it is only used on some variants).

How is selecting a driver a hw property within Zephyr?

**Comment by yashi** [comment]
> I disagree that the deferred initialization is strictly a software property. My main use case for this is to select the correct driver to initialize at runtime, which allows the same firmware image to be used on boards with multiple part stuffing options. I also have use cases where a board needs 2 instances of a driver, but one instance should automatically be initialized (it is present on all board variants) but the 2nd instance requires runtime check (it is only used on some variants).

Even with the use cases above, deferred initialization is still not solely a hardware concern. I'm not saying it's strictly a software thing either.

In my opinion, it's a system-level consideration. Our users should define their system, not us.

For example, consider two boards with exactly the same hardware components. One might be used in a system where deferred initialization is necessary. The other might require one instance of a device to be initialized automatically, while another instance of the same device needs to be initialized later. Would you want to prepare different device tree files for each scenario? In an ideal world, that might work, but I'm sure you would also need to modify your software accordingly.

**Comment by edersondisouza** [comment]
@tbursztyka do you have more comment about this approach vs #67877? Did you have a chance to take a look at the more recent discussion at #39896?

**Comment by edersondisouza** [comment]
And anyone else, really^ =D

**Comment by nashif** [comment]
no need for a DNM here, it is still under review and will not merged before getting the required reviews, so to avoid confusion, removing the label.

**Review by cfriedt — DISMISSED** [review]

**Comment by cfriedt** [comment]
I don't have any major complaints - lots of corner cases that I can think of that will need work, but no blockers.

Is there a way it can be marked experimental?

**Comment by fabiobaltieri** [comment]
@cfriedt weren't you suggesting for a callback? https://github.com/zephyrproject-rtos/zephyr/pull/67335#issuecomment-1887115782

**Comment by cfriedt** [comment]
It doesn't actually need to be a callback. It could be compile-time const, like a static delay added in a macro. But of course it could also be something as simple as returning `-EAGAIN`.

There are non-ideal situations where deferrals can cause other devices to fail to load.

We still need deinit. What about dynamic devices? Maybe non-malloced dynamic devices (e.g. possibly using devmux)?

I just think there is a lot of work to do here and it doesn't help to _defer_ that indefinitely.

**Review by nashif — DISMISSED** [review]

**Comment by tbursztyka** [comment]
> @tbursztyka, @cfriedt do you have more comment about this approach vs #67877? Did you have a chance to take a look at the more recent discussion at #39896?

I still think it's not the right way. Whether it's a callback and/or a dts property (which is worse than the cb imo), because of the dependencies, because of shortcomings in the device driver model. But I won't block, there seems to be a consensus already.

**Review by jhedberg — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

File: dts/bindings/base/base.yaml [hunk 5]
```diff
@@ -92,3 +92,9 @@ properties:
   mbox-names:
     type: string-array
     description: Provided names of mailbox / IPM channel specifiers
+
+  zephyr,deferred-init:
+    type: boolean
+    description: |
+      Do not initialize device automatically on boot. Device should be manually
+      initialized using z_device_init().
```
Comment (henrikbrixandersen) [inline dts/bindings/base/base.yaml:100]:
```suggestion
      initialized using device_init().
```

File: include/zephyr/device.h [hunk 6]
```diff
@@ -756,6 +766,22 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 	return z_device_is_ready(dev);
 }
 
+/**
+ * @brief Initialize a device.
+ *
+ * A device whose initialization was deferred (by marking it as
+ * `zephyr,deferred-init` on devicetree) needs to be initialized manually via
```
Comment (henrikbrixandersen) [inline include/zephyr/device.h:773]:
```suggestion
 * ``zephyr,deferred-init`` on devicetree) needs to be initialized manually via
```

File: tests/kernel/device/boards/hifive_unmatched.overlay
Comment (henrikbrixandersen) [inline tests/kernel/device/boards/hifive_unmatched.overlay:1]: Why add an overlay for this specific board?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
Architecture WG:

- Consensus is moving into the direction of using this approach
- No current objections to this approach

**Comment by edersondisouza** [comment]
v4:
  - Fix typo;
  - Fix documentation quotes;
  - Rebased.

**Review by nashif — APPROVED** [review]

File: tests/kernel/device/boards/hifive_unmatched.overlay
Comment (edersondisouza) [reply tests/kernel/device/boards/hifive_unmatched.overlay:1]: It was there already, I just updated it so tests build for this board =D
First patch that introduced them mentions:

> HiFive Unmatched is using size and address cells of length 2. It has to use different overlays (with reg properties of correct length).

**Review by gmarull — APPROVED** [review]
Not entirely happy with the result, but let's move on.

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**